modules/test/playwright/tests/portal-security-content-security-policy/contentSecurityPolicy.spec.ts

```diff
@@ -5,40 +5,822 @@
+
+	const DEFAULT_VIRTUAL_INSTANCE_NAME = 'www.able.com';
+
+	await pageEditorPage.editHTMLEditable({
+		editableId: 'element-html',
+		fragmentId: await pageEditorPage.getFragmentId('HTML'),
+		value: `<object
+					type="text/html"
+					data="http://${DEFAULT_VIRTUAL_INSTANCE_NAME}:8080"
+					width="300"
+					height="200">
+				</object>
+
+				<embed
+					type="text/html"
+					src="http://${DEFAULT_VIRTUAL_INSTANCE_NAME}:8080"
+					width="300"
+					height="200" />
+
+				<iframe
+					id="inlineFrameExample"
+					title="Inline Frame Example"
+					width="300"
+					height="200"
+					src="http://${DEFAULT_VIRTUAL_INSTANCE_NAME}:8080">
+				</iframe>`,
+	});
+
+	await pageEditorPage.publishPage();
+
+	await virtualInstancesPage.addNewVirtualInstance(
+		DEFAULT_VIRTUAL_INSTANCE_NAME
+	);
+
+	const defaultBaseUrl = liferayConfig.environment.baseUrl;
+
+	liferayConfig.environment.baseUrl = `http://${DEFAULT_VIRTUAL_INSTANCE_NAME}:8080`;
+
+	const newPage = await browser.newPage({
+		baseURL: `http://${DEFAULT_VIRTUAL_INSTANCE_NAME}:8080`,
+	});
+
+	await performLoginViaApi(
+		newPage,
+		'test',
+		`@${DEFAULT_VIRTUAL_INSTANCE_NAME}.com`
+	);
+
+	const virtualInstanceContentSecurityPolicyPage =
+		new ContentSecurityPolicyPage(newPage);
+
+	await test.step('CSP frame-ancestors allows framing from specific domain', async () => {
+		await virtualInstanceContentSecurityPolicyPage.gotoAndConfigurePolicy(
+			`frame-ancestors 'self' ${defaultBaseUrl};`
+		);
+
+		liferayConfig.environment.baseUrl = defaultBaseUrl;
+
+		await expect(async () => {
+			const cspErrors = [];
+
+			page.on('console', (msg) => {
+				if (
+					msg.type() === 'error' &&
+					msg
+						.text()
+						.includes(
+							'Content Security Policy directive: "frame-ancestors'
+						)
+				) {
+					cspErrors.push({text: msg.text(), type: msg.type()});
+				}
+			});
+
+			await page.goto(`/web/${site.name}/${layout.friendlyUrlPath}`);
+
+			await page.waitForLoadState();
+
+			expect(cspErrors.length).toBeGreaterThanOrEqual(0);
+		}).toPass();
+	});
+
+	await test.step('CSP frame-ancestors blocks framing from specific domain', async () => {
+		liferayConfig.environment.baseUrl = `http://${DEFAULT_VIRTUAL_INSTANCE_NAME}:8080`;
+
+		await virtualInstanceContentSecurityPolicyPage.gotoAndConfigurePolicy(
+			`frame-ancestors 'self';`
+		);
+
+		liferayConfig.environment.baseUrl = defaultBaseUrl;
+
+		await expect(async () => {
+			const cspErrors = [];
+
+			page.on('console', (msg) => {
+				if (
+					msg.type() === 'error' &&
+					msg
+						.text()
+						.includes(
+							'Content Security Policy directive: "frame-ancestors'
+						)
+				) {
+					cspErrors.push({text: msg.text(), type: msg.type()});
+				}
+			});
+
+			await page.goto(`/web/${site.name}/${layout.friendlyUrlPath}`);
+
+			await page.waitForLoadState();
+
+			expect(cspErrors.length).toBeGreaterThanOrEqual(3);
+		}).toPass();
+	});
+
+	await virtualInstancesPage.deleteVirtualInstance(
```

Comment:
If this test fails (even with a false-positive) after virtual instance creation but before deletion, the virtual instance will persist between tests.  If this would cause the next test to fail, then we will need to make sure the virtual instance is always removed between tests, likely with a `test.afterEach()` method.

If the test can run completely fine with the virtual instance already existing, then we are fine and this comment can be marked as resolved.